File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.py
class Solution:
    '''
    - if s_num is in array, ie if len(arr) + 1 >= s_num, return arr[s_num]
    - else
    '''
    def tribonacci(self, n: int) -> int:
        arr = [0, 1, 1]
        if n == 0: return 0
        if n == 1 or n == 2: return arr[1] or arr[2]
        
        def rec(s_num):
            if len(arr) > s_num:
                return arr[s_num]
            else:
                temp = rec(s_num -3) + rec(s_num -2) + rec(s_num - 1)
                arr.append(temp)
                return temp
            
                      
        return rec(n)
    
    
    
        
        
        
        
            
        
        

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.ts
function tribonacci(n: number): number {
    let arr = [0, 1, 1];
    if(n == 0) return arr[0];
    if( n == 1 || n == 2) return arr[1];
    
    const rec = (num) => {
        if (arr.length > num){
            return arr[num]
        } else{
            let temp = rec(num - 3) + rec(num - 2) + rec(num - 1)
            arr[num] = temp;
            return temp;
        }
    }
    
    return rec(n)
};